[FEAT] 리뷰 작성, 수정, 삭제 기능 개발

jd99iam opened the pull request:
## #118 

### 💡 PR 타입(하나 이상의 PR 타입을 선택해주세요)
* [x] 기능 추가
* [x] 테스트코드 작성

### 📲 반영 브랜치
ex) feat/118-write-review -> 1-develop

### ❔ 변경 사항
리뷰 작성 기능을 추가했습니다
리뷰 수정 기능을 추가했습니다
리뷰 삭제 기능을 추가했습니다

### 🔍 테스트 결과
- Service 테스트
내 리뷰 아닌 리뷰 수정, 삭제 시 예외 발생 테스트 추가
이미 삭제된 리뷰 수정, 삭제 시 예외 발생 테스트 추가

### 📰 참고 Reference

### ⚒️ 기타 사항

jd99iam committed:
feat: ReviewController 리뷰 작성 추가

Related to : #118

jd99iam committed:
feat: ReviewController 리뷰 작성 추가

Related to : #118

jd99iam committed:
feat: ReviewPostRequest 추가

Related to : #118

jd99iam committed:
feat: ReviewPostRequest에 정적 팩토리 메서드 추가

Related to : #118

jd99iam committed:
feat: writeReview 인자 추가

Related to : #118

jd99iam committed:
feat: findSolutionById 메서드 추출

Related to : #118

jd99iam committed:
feat: ReviewNotFoundException 추가

Related to : #118

jd99iam committed:
feat: ReviewPatchRequest 추가

Related to : #118

jd99iam committed:
feat: ReviewController 리뷰 수정 추가

Related to : #118

jd99iam committed:
feat: Review 수정 기능 추가

Related to : #118

jd99iam committed:
feat: Review 패키지 이동

Review 패키지 이동
ReviewStatusConstant 추가
Review Entity에 상태 필드 추가

Related to : #118

jd99iam committed:
feat: Review update 메서드 수정

상태도 변경되도록 수정

Related to : #118

jd99iam committed:
feat: Review 생성 메서드 수정

Review 생성 시 상태 CREATED 되도록 변경

Related to : #118

jd99iam committed:
feat: Review 응답 dto 변경

응답 dto에서 상태도 응답하도록 변경

Related to : #118

jd99iam committed:
feat: Valid 메세지 추가

Related to : #118

jd99iam committed:
feat: ReviewController 리뷰 삭제 메서드 추가

Related to : #118

jd99iam committed:
feat: ReviewService 리뷰 삭제 기능 추가

Related to : #118

jd99iam committed:
feat: 이미 삭제된 리뷰에 대한 예외 추가

1. 이미 삭제된 리뷰를 수정하려고 하면 예외 발생
2. 이미 삭제된 예외를 삭제하려고 하면 예외 발생

Related to : #118

jd99iam committed:
feat: ReviewFixture 추가

Related to : #118

jd99iam committed:
feat: ReviewPatchRequest에 생성자 추가

Related to : #118

jd99iam committed:
feat: FollowRepositoryTest에 필드명 변경

Related to : #118

jd99iam committed:
test: 리뷰 테스트 추가

1. 자신의 리뷰가 아닌 리뷰를 수정, 삭제하려 할 시 예외 발생
2. 이미 삭제된 리뷰를 수정, 삭제하려 할 시 예외 발생

Related to : #118

jd99iam committed:
Merge branch '1-develop' of https://github.com/rlfrkdms1/Prog-rangers into feat/118-write-review

jd99iam committed:
feat: 풀이 상세보기 풀이 줄바꿈

Related to : #118

jd99iam committed:
test: 불필요한 주석 삭제

Related to : #118

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/controller/ReviewController.java:
```diff
@@ -1,25 +1,47 @@
+        return redirect(solutionId);
+    }
+
+    private ResponseEntity<Object> redirect(Long solutionId) {
```
왜 ResponseEntity<Object>를 반환하나요 ? 그냥 ResponseEntity<Void>를 반환하면 안되는걸까요 ?

jd99iam replied:
> 왜 ResponseEntity를 반환하나요 ? 그냥 ResponseEntity를 반환하면 안되는걸까요 ?

수정했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/dto/review/request/ReviewPostRequest.java:
```diff
@@ -0,0 +1,19 @@
+@AllArgsConstructor
+@NoArgsConstructor
+public class ReviewPostRequest {
+    @NotNull
```
int인데 `NotNull`을 사용하면 적용이 안될 것 같아요

jd99iam replied:
> int인데 `NotNull`을 사용하면 적용이 안될 것 같아요

수정했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/ReviewService.java:
```diff
@@ -38,6 +44,11 @@ public SolutionReviewsResponse getReviewDetail(Long solutionId) {
     }
 
+    private Solution findSolutionById(Long solutionId) {
+        Solution solution = solutionRepository.findById(solutionId).orElseThrow(SolutionNotFoundException::new);
```
줄을 합칠 수 있을 것 같아요

jd99iam replied:
> 줄을 합칠 수 있을 것 같아요

수정했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/ReviewService.java:
```diff
@@ -74,4 +85,14 @@ private static void makeReplyResponse(List<SolutionReview> solutionReviewRespons
         }
     }
+
+    public void writeReview(ReviewPostRequest reviewPostRequest, Long memberId,Long solutionId) {
```
service클래스에 `@Transactional` 빠졌어요

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/ReviewService.java:
```diff
@@ -74,4 +118,25 @@ private static void makeReplyResponse(List<SolutionReview> solutionReviewRespons
+        }
+    }
+
+    private static void checkReviewAlreadyDeleted(Review targetReview) {
```
저희 다 `valid`로 통일했던 것 같아요

jd99iam replied:
> 저희 다 `valid`로 통일했던 것 같아요

수정했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/ReviewService.java:
```diff
@@ -74,4 +118,25 @@ private static void makeReplyResponse(List<SolutionReview> solutionReviewRespons
+        return reviewRepository.findById(reviewId).orElseThrow(ReviewNotFoundException::new);
+    }
+
+    private static void checkMemberAuthorization(Review targetReview, Member member) {
```
`static`이 왜 사용된 걸까요 ?

jd99iam replied:
> `static`이 왜 사용된 걸까요 ?

불필요한 static 삭제했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/ReviewService.java:
```diff
@@ -1,32 +1,71 @@
+    }
+
+    @Transactional
+    public void deleteReview(Long memberId, Long reviewId) {
```
여기는 `review`가 여러개 사용되지 않았는데 왜 targetReview인가요 ?

jd99iam replied:
> 여기는 `review`가 여러개 사용되지 않았는데 왜 targetReview인가요 ?

수정했습니다

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/entity/review/Review.java:
```diff
@@ -47,4 +52,18 @@ public class Review {
+    }
+
+    public void delete() {
+        this.status=DELETED;
```
상태가 status로 변경될 때 content 내용도 바뀌어야 되지 않을까요 ?

jd99iam replied:
> 상태가 status로 변경될 때 content 내용도 바뀌어야 되지 않을까요 ?

리뷰나 댓글 내용은 놔두고 status로 관리하기로 한 걸로 이해했는데 아닌가요?

rlfrkdms1 replied:
`삭제된 댓글입니다`라고 뜨지 않고 그냥 그대로 두는건가요 ? 그렇다면 삭제의 이유가 있을까요 ?

jd99iam replied:
> `삭제된 댓글입니다`라고 뜨지 않고 그냥 그대로 두는건가요 ? 그렇다면 삭제의 이유가 있을까요 ?

전에 content를 `삭제된 댓글입니다` 로 뜨게 하면 사용자가 댓글을 `삭제된 댓글입니다` 로 입력한 것이랑 구분이 안 된다고 하시지 않으셨나요?

rlfrkdms1 replied:
content를 `삭제된 댓글입니다`로만 고치면 구분이 안되서 status를 추가하자고 말씀드렸던 겁니다 !

jd99iam replied:
> content를 `삭제된 댓글입니다`로만 고치면 구분이 안되서 status를 추가하자고 말씀드렸던 겁니다 !

이해했습니다! content도 수정되도록 바꿨습니다. 
그렇다면 프론트 분들과 상의해서 `STATUS = DELETED` 인 경우 클라이언트에게 뭔가 다르게 보여지도록 해야 STATUS를 넣은 의미가 있을 것 같아요!

rlfrkdms1 replied:
좋습니다 ~ 회의 안건에 적어놓도록 하겠습니다

rlfrkdms1 commented on backend/src/test/java/com/prograngers/backend/service/ReviewServiceTest.java:
```diff
@@ -0,0 +1,103 @@
+        Member member1 = 장지담.아이디_지정_생성(1L);
+        Problem problem = 백준_문제.기본_정보_생성();
+        // 스크랩 당할 풀이 scrapTarget
+        Solution solution = 공개_풀이.아이디_지정_생성(1L,problem,member1, LocalDateTime.now(),DFS,LIST,JAVA,1);
```
solution의 알고리즘, 자료구조가 필수 항목이 아닌데, 이들이 필요없는 테스트에서 사용될 필요가 있을까요 ?

jd99iam replied:
> solution의 알고리즘, 자료구조가 필수 항목이 아닌데, 이들이 필요없는 테스트에서 사용될 필요가 있을까요 ?

알고리즘, 자료구조 필수 항목입니다

rlfrkdms1 replied:
init.sql 보면 필수 항목 아니라고 지정하기로 했는데, 필수항목이라고 하신 근거가 뭔가요 ?

jd99iam replied:
> init.sql 보면 필수 항목 아니라고 지정하기로 했는데, 필수항목이라고 하신 근거가 뭔가요 ?

풀이 작성할 때 알고리즘, 자료구조는 필수로 있어야 하지 않나요? init.sql을 수정해야 할 것 같습니다

rlfrkdms1 replied:
init.sql 은 저희가 같이 수정하고 만든건데, 왜 필수가 아니라고 결론지었을지 생각해 보면 좋을 것 같아요.
queue를 사용한 경우엔 자료구조만 사용한 경우인데, 알고리즘은 넣을게 없는 경우가 생기지 않을까요 ?

jd99iam replied:
> init.sql 은 저희가 같이 수정하고 만든건데, 왜 필수가 아니라고 결론지었을지 생각해 보면 좋을 것 같아요. queue를 사용한 경우엔 자료구조만 사용한 경우인데, 알고리즘은 넣을게 없는 경우가 생기지 않을까요 ?

말씀 들어보니 맞는 것 같습니다.
그렇다면 풀이 생성할 때 알고리즘, 자료구조 둘 다 null일수도 있는 걸까요?
사용자가 정한 알고리즘, 자료구조 중에 특별히 사용한 게 없다고 생각할 수도 있지 않을까요?

rlfrkdms1 replied:
그런 경우를 방지하려면 service 단에서 검증하면 될 것 같습니다 DB 단에서 검증할 방법은 없을 것 같아요

rlfrkdms1 replied:
음 둘 다 null인 경우도 허용해야된다는 말씀이실까요 ?
저는 사실 둘 다 null인 경우도 허용해야 된다는 입장입니다.
태그는 분류를 통한 편의를 제공하는 기능이지 필수 기능은 아니라고 생각해서요

jd99iam replied:
> 음 둘 다 null인 경우도 허용해야된다는 말씀이실까요 ? 저는 사실 둘 다 null인 경우도 허용해야 된다는 입장입니다. 태그는 분류를 통한 편의를 제공하는 기능이지 필수 기능은 아니라고 생각해서요

넵 알겠습니다

jd99iam replied:
> > 음 둘 다 null인 경우도 허용해야된다는 말씀이실까요 ? 저는 사실 둘 다 null인 경우도 허용해야 된다는 입장입니다. 태그는 분류를 통한 편의를 제공하는 기능이지 필수 기능은 아니라고 생각해서요
> 
> 넵 알겠습니다

픽스처 수정해서 알고리즘, 자료구조 없이 사용할 수 있도록 하겠습니다

rlfrkdms1 replied:
제가 올린 것 중에 있습니다 사용하시면 될 것 같아요 !

jd99iam replied:
> 제가 올린 것 중에 있습니다 사용하시면 될 것 같아요 !

그럼 merge 후에 수정하도록 하겠습니다

rlfrkdms1 commented on backend/src/test/java/com/prograngers/backend/service/ReviewServiceTest.java:
```diff
@@ -0,0 +1,103 @@
+                ()->assertThrows(ReviewAlreadyDeletedException.class,()->reviewService.deleteReview(1L,1L))
+        );
+    }
+    ReviewPatchRequest 리뷰_수정_요청_생성(){
```
접근 제한자가 누락된 것 같네요,

jd99iam replied:
> 접근 제한자가 누락된 것 같네요,

추가했습니다

jd99iam committed:
refactor: 컨트롤러에서 redirect 메서드 ResponseEntity 반한하도록 수정

Related to : #118

jd99iam committed:
fix: NotNull이 적용되도록 codeLineNumber 자료형 수정

Related to : #118

jd99iam committed:
refactor: 라인 합침

Related to : #118

jd99iam committed:
refactor: check valid로 수정

Related to : #118

jd99iam committed:
refactor: 불필요한 static 제거

Related to : #118

jd99iam committed:
refactor: targetReview -> review로 작명 변경

Related to : #118

jd99iam committed:
refactor: 접근 제한자 추가

Related to : #118

jd99iam committed:
Merge branch '1-develop' into feat/118-write-review

jd99iam committed:
Merge remote-tracking branch 'upstream/feat/118-write-review' into feat/118-write-review

# Conflicts:
#	backend/src/main/java/com/prograngers/backend/controller/LikesController.java
#	backend/src/main/java/com/prograngers/backend/controller/SolutionController.java
#	backend/src/main/java/com/prograngers/backend/service/ProblemService.java
#	backend/src/main/java/com/prograngers/backend/service/SolutionService.java

jd99iam committed:
fix: 충돌 해결

Related to : #118

jd99iam committed:
refactor: ResponseEntity 반환하도록 수정

Related to : #118

jd99iam committed:
fix: 좋아요 누르기, 취소 매핑 변경

Related to : #118

jd99iam committed:
fix: ProblemService 에러 수정

1. stream.toList로 생성한 리스트를 사용해서 예외 발생, new ArrayList로 생성하는 방식으로 변경
2. tagCount랑 비교하는 로직 변경

Related to : #118

jd99iam committed:
fix: 댓글, 리뷰 삭제 시 content 변경

삭제된 댓글,리뷰 입니다로 content 변경

Related to : #118

jd99iam committed:
refactor: checkReviewIsMine 로직 위치 이동

dto에서 서비스로 이동

Related to : #118

jd99iam committed:
refactor: checkReviewIsMine 로직 위치 이동

dto에서 서비스로 이동

Related to : #118

jd99iam committed:
refactor: checkLink 리팩터링

1. dto에서 서비스로 로직 이동
2. valid로 rename

Related to : #118

jd99iam committed:
refactor: checkReviewIsMine rename

validReviewIsMine으로 rename

Related to : #118

jd99iam committed:
refactor: 검증 로직 이동

1. 검증 로직 dto에서 서비스로 이동
2. 서비스에서 dto를 각각 생성 후 합쳐서 반환하도록 변경

Related to : #118

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -87,44 +94,101 @@ public void delete(Long solutionId, Long memberId){
+            List<SolutionDetailComment> replies = commentResponse.getReplies();
+            if (replies!=null){
+                for (SolutionDetailComment commentReplyResponse : replies){
+                    if (commentReplyResponse.getNickname().equals(member.getNickname())) commentReplyResponse.setMine(true);
```
`setMine` 말고 다른 작명이 좋을 것 같아요.
DTO 단에서의 setter 사용은 자유롭게 사용하는걸까요 ?

jd99iam replied:
> `setMine` 말고 다른 작명이 좋을 것 같아요. DTO 단에서의 setter 사용은 자유롭게 사용하는걸까요 ?

DTO에서의 setter 사용은 상관 없다고 하시지 않으셨나요?

rlfrkdms1 replied:
제가 그렇게 말씀 드렸나요 ?_? 알겠습니당

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -87,44 +94,101 @@ public void delete(Long solutionId, Long memberId){
+        return SolutionDetailResponse.from(solutionDetailProblem,solutionDetailSolution,solutionDetailComments);
+    }
+
+    private void setCommentMine(Member member, List<SolutionDetailComment> solutionDetailComments) {
```
setCommentMine을 따로 하지 말고 댓글 response를 만들때 같이 검증할 순 없나요 ?

jd99iam replied:
> setCommentMine을 따로 하지 말고 댓글 response를 만들때 같이 검증할 순 없나요 ?

검증 로직을 서비스 단에서 하자고 하셔서 이렇게 수정한 것입니다..

rlfrkdms1 replied:
밑에 서비스 단에서 댓글 response를 만들어서 넣고 있지 않나요 ?

jd99iam replied:
> 밑에 서비스 단에서 댓글 response를 만들어서 넣고 있지 않나요 ?

수정했습니다!!

jd99iam committed:
refacotr: 내 댓글인지 검증하는 로직을 댓글 response 만드는 안으로 넣기

Related to : #118

jd99iam committed:
refacotr: 내 댓글인지 검증하는 로직을 댓글 response 만드는 안으로 넣기

Related to : #118

rlfrkdms1 commented on backend/src/main/java/com/prograngers/backend/service/SolutionService.java:
```diff
@@ -167,28 +166,43 @@ private String getScrapSolutionLink(Solution solution){
+            commentResponseList.add(SolutionDetailComment.from(comment, new ArrayList<>(), true));
+            return;
+        }
+            commentResponseList.add(SolutionDetailComment.from(comment, new ArrayList<>(), false));
```
if 문 안쓰고 그냥 `commentResponseList.add(SolutionDetailComment.from(comment, new ArrayList<>(), validCommentIsMine(memberId, comment)));`로 변경하면 될 것 같은데요 ?
그리고 valid 말고 그냥 IsCommentMine이나 IsMine등으로 변경하는게 좋을 것 같습니다. valid는 검증할 때 사용하는 건데 해당 메서드는 검증과는 거리가 멀어보여요

jd99iam replied:
> if 문 안쓰고 그냥 `commentResponseList.add(SolutionDetailComment.from(comment, new ArrayList<>(), validCommentIsMine(memberId, comment)));`로 변경하면 될 것 같은데요 ? 그리고 valid 말고 그냥 IsCommentMine이나 IsMine등으로 변경하는게 좋을 것 같습니다. valid는 검증할 때 사용하는 건데 해당 메서드는 검증과는 거리가 멀어보여요

수정했습니다!

jd99iam committed:
refacotr: 불필요한 if문 삭제, 메서드 rename

1. 댓글이 내 댓글인지 확인하는 과정에서 불필요한 if문 삭제
2. 메서드 명 validCommentIsMine에서 checkCommmentIsMine으로 rename

Related to : #118
